### PR TITLE
rust: use lld on pkgsStatic aarch64

### LINKED ISSUE #254454: Build failure: pkgsStatic.fd on aarch64-linux
### Steps To Reproduce

1. build pkgsStatic.fd on aarch64-linux

### Build log

https://termbin.com/x4lm

### Notify maintainers

unclear, rust team? @figsoda @Mic92 @tjni @winterqt @zowoq 

### Additional Context

https://github.com/rust-lang/rust/issues/89626

### Metadata

latest master

### PR DESCRIPTION
**Pull request by yu-re-ka**
## Description of changes

Fixes https://github.com/NixOS/nixpkgs/issues/254454 https://github.com/NixOS/nixpkgs/issues/264687

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by alyssais** [comment]
Can't this target master?

**Comment by yu-re-ka** [comment]
Indeed it can

File: pkgs/build-support/rust/lib/default.nix [hunk 1]
```diff
@@ -12,17 +12,25 @@ rec {
   # hostPlatform-targeted compiler -- for example, `-m64` being
   # passed on a build=x86_64/host=aarch64 compilation.
   envVars = let
+
+    isAarch64Static = platform: platform.isAarch64 && platform.isMusl && platform.isStatic;
```
Comment (alyssais) [inline pkgs/build-support/rust/lib/default.nix:16]: It's _probably_ not musl-specific, right?  The issue is just static libraries not having dependency information, as I understand it.  But we wouldn't want to link with lld on Darwin, so maybe

```suggestion
    shouldUseLLD = platform: platform.isAarch64 && platform.isStatic && !platform.isDarwin;
```

I also think we need some explanation of why we're using lld, I guess with a link to the issue.

File: pkgs/build-support/rust/lib/default.nix [hunk 2]
```diff
@@ -12,17 +12,25 @@ rec {
...
+
     ccForBuild = "${buildPackages.stdenv.cc}/bin/${buildPackages.stdenv.cc.targetPrefix}cc";
     cxxForBuild = "${buildPackages.stdenv.cc}/bin/${buildPackages.stdenv.cc.targetPrefix}c++";
+    linkerForBuild = ccForBuild;
+
     ccForHost = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}cc";
     cxxForHost = "${stdenv.cc}/bin/${stdenv.cc.targetPrefix}c++";
+    linkerForHost = if isAarch64Static stdenv.targetPlatform then "${buildPackages.lld}/bin/ld.lld" else ccForHost;
```
Comment (alyssais) [inline pkgs/build-support/rust/lib/default.nix:24]: We should check whether stdenv's linker is lld already, so we don't unnecessarily override it with a potentially different version.  Something like

```suggestion
    linkerForHost = if isAarch64Static stdenv.targetPlatform && !stdenv.cc.bintools.isLLVM then "${buildPackages.lld}/bin/ld.lld" else ccForHost;
```

**Review by alyssais — CHANGES_REQUESTED** [review]

File: pkgs/build-support/rust/lib/default.nix (hunk 2 above)
Comment (alyssais) [reply pkgs/build-support/rust/lib/default.nix]: (And also for the target.)

**Comment by yu-re-ka** [comment]
I have added the requested changes

**Review by alyssais — APPROVED** [review]

**Pull request merged**